### PR TITLE
Clarify that an appservice can be interested in local and remote users

### PR DESCRIPTION
**Pull request by MadLittleMods**
Clarify that an appservice can be interested in local and remote users. Also adds tests that would actually fail if we attempted something like https://github.com/matrix-org/synapse/pull/13958 again.

---

> It's not really a question of controlling the users, but more of whether the application service wants to receive events and updates related to those users, which an application service may well want to do for events coming from non-local users.
>
> *-- @anoadragon453, https://github.com/matrix-org/synapse/pull/13958#discussion_r984535807*

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: tests/handlers/test_appservice.py [hunk 1]
```diff
@@ -412,6 +413,146 @@ def prepare(self, reactor, clock, hs):
...
+        # This is normally set in `notify_interested_services` but we need to call the
+        # internal async version so the reactor gets pushed to completion.
+        self.hs.get_application_service_handler().current_max += 1
+        self.get_success(
+            self.hs.get_application_service_handler()._notify_interested_services(
+                RoomStreamToken(
+                    None, self.hs.get_application_service_handler().current_max
+                )
+            )
+        )
```
Comment (MadLittleMods) [inline tests/handlers/test_appservice.py:426]: Is there a better, more clean way to do this dance?

File: tests/handlers/test_appservice.py [hunk 2]
```diff
@@ -412,6 +413,146 @@ def prepare(self, reactor, clock, hs):
...
+            _device_list_summary,
+        ) = self.send_mock.call_args[0]
+
+        # Even though the message came from an uninsteresting user, it should still
+        # notify us because the interesting user is joined to the room.
+        self.assertEqual(service, interested_appservice)
+        self.assertEqual(events[0]["type"], "m.room.message")
+        self.assertEqual(events[0]["sender"], alice)
+
+    def test_match_remote_room_members(self):
```
Comment (MadLittleMods) [inline tests/handlers/test_appservice.py:492]: These tests are practically the same. Only difference is the user namespace regex and the interesting user itself.

We could change this to a for-loop:

```py
for interesting_user in ["@local_as_user:test", "@interesting_user:remote"]:
    def test_match_interesting_room_members(self):
       ...
```

Comment (MadLittleMods) [reply tests/handlers/test_appservice.py]: Done using `@parameterized.expand(["@local_as_user:test", "@interesting_user:remote"])`

File: tests/handlers/test_appservice.py [hunk 3]
```diff
@@ -412,6 +413,83 @@ def prepare(self, reactor, clock, hs):
...
+            service,
+            events,
+            _ephemeral,
+            _to_device_messages,
+            _otks,
+            _fbks,
+            _device_list_summary,
+        ) = self.send_mock.call_args[0]
+
+        # Even though the message came from an uninsteresting user, it should still
```
Comment (anoadragon453) [inline tests/handlers/test_appservice.py:487]:
```suggestion
        # Even though the message came from an uninteresting user, it should still
```

**Review by anoadragon453 — COMMENTED** [review]
I don't have time at the moment for a full review, but with an AS hat on the comments and code look correct to me.

**Comment by erikjohnston** [comment]
> It's not really a question of controlling the users, but more of whether the application service wants to receive events and updates related to those users, which an application service may well want to do for events coming from non-local users.

Is there a legitimate use case for this though? I'm a bit wary as the behaviour will be different based on local or remote users. In particular, you'll only get stuff from remote users in rooms that the server is in, at which point you'll usually get that info as they'll be AS rooms.

I just want to make sure that this is something we consciously want to support, rather than something that someone might want at some point, maybe.

**Comment by anoadragon453** [comment]
Thanks for questioning this @erikjohnston. After thinking about this more and [asking in #matrix-spec](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$c9mpH1UM80sEewLzXTQtpTGjeQYBkmd2M-gnvNnGrxQ?via=matrix.org&via=libera.chat&via=element.io), it indeed seems that there are *no valid use cases* for needing to specify a remote user in an application service's user namespace. By specifying local users, you will receive room events/presence updates from remote users anyhow.

As @Half-Shot put it, defining remote users is merely a footgun for which an Application Service may assume that it'll receive all events sent by that user, even though it will only receive events in rooms that are shared with a local user.

So by that logic, I'm fine with us defining (and limiting in the code) user namespaces to only cover local users. It would make sense to update the spec to include the same. Sorry for the run-around!

**Comment by MadLittleMods** [comment]
Sounds great to me 👍

I'll get https://github.com/matrix-org/synapse/pull/13958 in shape again ⏩

File: tests/handlers/test_appservice.py [hunk 4]
```diff
@@ -412,6 +413,83 @@ def prepare(self, reactor, clock, hs):
...
+                )
+            )
+        )
+
+    @parameterized.expand(["@local_as_user:test", "@interesting_user:remote"])
+    def test_match_interesting_room_members(self, interesting_user):
+        """
+        Test to make sure that a interesting user (local or remote) in the room is notified
+        when someone else in the room sends a message.
+        """
```
Comment (MadLittleMods) [inline tests/handlers/test_appservice.py:433]: TODO: Potentially merge these tests into https://github.com/matrix-org/synapse/pull/13958

Per-discussion, https://github.com/matrix-org/synapse/pull/14000#issuecomment-1266552433, they need to be adapted to not expect a remote user anymore though

Comment (MadLittleMods) [reply tests/handlers/test_appservice.py:433]: Adapted and moved over to https://github.com/matrix-org/synapse/pull/13958 ✅

**Comment by MadLittleMods** [comment]
> It would make sense to update the spec to include the same.

Also created https://github.com/matrix-org/matrix-spec/issues/1272 to track a spec clarification ⏩

**Pull request closed without merging**